src/core/net/dns_headers.hpp

```diff
@@ -2340,20 +2329,22 @@ class OptRecord : public ResourceRecord
     //  +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
     //  |         EXTENDED-RCODE        |            VERSION            |
     //  +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
-    //  | DO|                           Z                               |
+    //  | DO|                Z          |             Z                 |
     //  +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
     //
     // The variable data part of OPT RR can contain zero of more `Option`.
 
-    enum : uint32_t
+    enum : uint8_t
     {
-        kExtRCodeffset = 24,                    // Extended RCODE field offset.
-        kExtRCodeMask  = 0xf << kExtRCodeffset, // Extended RCODE field mask.
-        kVersionOffset = 16,                    // Version field offset.
-        kVersionMask   = 0xf << kVersionOffset, // Version field mask.
-        kDnsSecFlag    = 1 << 15,               // DnsSec bit flag.
+        kExtRCodeByteIndex = 0,      // Byte index of Extended RCODE within the TTL field.
+        kVersionByteIndex  = 1,      // Byte index of Version within the TTL field.
+        kFlagByteIndex     = 2,      // Byte index of flag byte within the TTL field.
+        kDnsSecFlag        = 1 << 7, // DNSSec OK bit flag.
     };
 
+    uint8_t  GetTtlByteAt(uint8_t aIndex) const { return reinterpret_cast<const uint8_t *>(&mTtl)[aIndex]; }
+    uint8_t &GetTtlByteAt(uint8_t aIndex) { return reinterpret_cast<uint8_t *>(&mTtl)[aIndex]; }
```

Comment:
It is a little bit weird to assign to the return value of a method, although it works. Instead of having a `GetTtl*` method returns the reference, what about using a  symmetrical `SetTtlByteAt` that sets the value?

Reply:
I actually had a `SetTtlByteAt()` method initially in the code but switched it to this model (of getting a referencee to the `uint8_t`). The main reason is the new model allows code like this:
```
GetTtlByteAt(kFlagByteIndex) &= ~kDnsSecFlag;
```
instead of
```
SetTtlByteAt(kFlagByteIndex, GetTtlByteAt(kFlagByteIndex) & ~kDnsSecFlag);
```
Since it's private I felt making it easier to read would be better.

------

In general, we do use the concept of returning a reference from a getter in OT core in different modules (instead of or in addition to a setter). 
- For example, in `SockAddr` we have a `GetAddress()` which returns a reference.
- The main reason for this is to allow caller to use methods of `Ip6::Address` directly (e.g. `sockAddr.GetAddress().Clear()` or `sockAddr.GetAddress().SetToRoutingLocator(...)`, etc).